### ISSUE TITLE
Low performance with big files

### ISSUE THREAD

**Issue by Shatur95** [hrsh7th/nvim-compe#321]
Steps to reproduce:

* Open very big file (1.5M lines)
* Switch to insert mode.
* Start typing.

Profiling says that this caused by the following function:

```
FUNCTION  <SNR>27_on_text_changed()
    Defined: ~/.config/nvim/pack/plugins/opt/nvim-compe/plugin/compe.vim:39
Called 36 times
Total time:   2.173349
 Self time:   2.171816

count  total (s)   self (s)
   36   2.173265   2.171732   call luaeval('require"compe"._on_text_changed()')
```

Suggestion: maybe add an option to disable completion for files that bigger than specified number of lines?

**Comment by tzachar** [comment]
Would be interesting to know which source specifically causes this.

Anyway, if you are using packer, you can use the following to only load a plugin if the number of lines is less than 1000:
```lua
inspected_buffers = {}
-- this function is called b4 a buffer is opened, so we need to manually open the file and count the lines.
function limit_by_line_count(max_lines)
	local fname = vim.fn.expand("%:p")
	local cache = inspected_buffers[fname]
	if cache ~= nil then
		return cache
	end

	max_lines = max_lines or 1000
	local lines = 0
	for _ in io.lines(fname) do
		lines = lines + 1
		if lines > max_lines then
			break
		end
	end
	inspected_buffers[fname] = (lines <= max_lines)
	return inspected_buffers[fname]
end
```
and then in packer:

```lua
		use {'hrsh7th/nvim-compe', cond = limit_by_line_count, opt = true}
```

**Comment by hrsh7th** [comment]
I think this slows down is occurred by `buffer` source.

**Comment by hrsh7th** [comment]
https://github.com/hrsh7th/nvim-compe/blob/master/lua/compe_buffer/buffer.lua#L21

**Comment by tzachar** [comment]
Why would the buffer source need to re-read the file every time the user goes into insert mode?
Maybe it would be better to read once on open, and on every write?

**Comment by Shougo** [comment]
It is possible, but it cannot collect your inputted words..

**Comment by tzachar** [comment]
I dont know.
For me, it does not make sense to read the entire buffer every time you go into insert mode.
This should be done opportunistly, both time-wise and buffer-location-wise. What I mean is that
(a) the entire buffer should be read once upon load, and every once in a while afterwards
(b) when the user exits insert mode, reprocess the changed region only.

**Comment by Shougo** [comment]
The entire buffer scan is very slow, but it is accurate behavior.
This is trade off.

**Comment by Shougo** [comment]
The buffer source can reprocess the changed region, but it is not accurate result.

**Comment by Shatur95** [comment]
> Anyway, if you are using packer, you can use the following to only load a plugin if the number of lines is less than 1000:

This method will not work if I first open a small file and then a large one.

> I think this slows down is occurred by buffer source.

Yes, this happens when only the `buffer` source is enabled.

I think it's nice to be able to completely disable auto-completion (e.g. all sources) for large files.

**Comment by hrsh7th** [comment]
The buffer source doesn't re-load on each InsertEnter.
It loads all words initially and then watches line changes.

**Comment by hrsh7th** [comment]
The all word indexing is too heavy so we should improve/delay/skip it.

**Comment by hrsh7th** [comment]
```viml
autocmd BufEnter * call s:on_buf_enter()
function! s:on_buf_enter() abort
  if nvim_buf_line_count(0) < 2000
    call compe#setup({
    \  ...
    \ }, 0)
  endif
endfunction
```

**Comment by tjdevries** [comment]
@hrsh7th maybe you can use `nvim_buf_attach` to only parse updated lines. First pass may be a bit slow, but afterwards should be very quick.

**Comment by WhyNotHugo** [comment]
@hrsh7th Regarding [this last comment of yours](https://github.com/hrsh7th/nvim-compe/issues/321#issuecomment-814789105), what do you put in that `...` block to disable compe?

**Comment by Shatur95** [comment]
>  It loads all words initially and then watches line changes.
> The all word indexing is too heavy so we should improve/delay/skip it.

But I forgot to clarify that the problem appears only typing. So I think that the problem is not in the indexing of all words.

**Comment by hrsh7th** [comment]
The indexing of all words is triggered by first typing. So I still think it is the cause.

**Comment by Shatur95** [comment]
> The indexing of all words is triggered by first typing. So I still think it is the cause.

Thanks for the clarification!

**Comment by Shatur95** [comment]
@hrsh7th, can we have a proposed workaround as an option?

**Comment by mortezadadgar** [comment]
occurs to me on every files if i open a new file this takes 1-10ms time for editor input to respond
and seems related to vsnip completion source

**Comment by Shatur** [comment]
>  and seems related to vsnip completion source

No, I can reproduce it with buffer completion too.

**Comment by mortezadadgar** [comment]
> occurs to me whenever I open a new file takes 1-10ms for editor input to respond
> and seems related to vsnip completion source

@hrsh7th do you have any idea about this issue or I should open a new issue in vsnip repository

**Comment by Shougo** [comment]
@mortezadadgar Can you test it without vsnip source or without vnip?

> No, I can reproduce it with buffer completion too.

It seems not vsnip related problem...
The minimal configuration is needed.

**Comment by mortezadadgar** [comment]
@Shougo Yeah I tested without vsnip source and the delay on first character insert vanishes
I can reproduce with this config: https://pastebin.com/F2A0NVwe

**Comment by Shougo** [comment]
Oh. So it seems vsnip source problem interesting.
The delay is the first character only?
It may be from cache creation.

**Comment by mortezadadgar** [comment]
> The delay is the first character only?
Yes

**Comment by Shougo** [comment]
I think vsnip can load snippets before input.  I don't know what function do it though.

**Comment by Shougo** [comment]
You can call vsnip#get_complete_items(bufnr('％')) before input instead.  I think it can be called in autocmd FileType.

**Comment by mortezadadgar** [comment]
> call vsnip#get_complete_items(bufnr('％'))

yes this improve the situation but It won't work if I lazy load on InsertCharPre event

**Comment by Shougo** [comment]
Yes.  Because the delay is lazy loaded overhead.
No magic.

**Comment by hrsh7th** [comment]
@mortezadadgar The work around is here.

`autocmd FileType * call vsnip#get_complete_items(bufnr('%'))`

**Comment by hrsh7th** [comment]
I've tried to improve buffer source performance by Luv's Thread and message-pack RPC.

But it is impossible with current neovim (because neovim does not expose mpack module)

**Comment by mortezadadgar** [comment]
> @mortezadadgar The work around is here.
> 
> `autocmd FileType * call vsnip#get_complete_items(bufnr('%'))`

Is there any better solution for this issue

**Comment by Shougo** [comment]
Unfortunately, no.  If you don't like the overhead, you cannot lazy vim-vsnip.

**Comment by mortezadadgar** [comment]
@hrsh7th sorry I'm spamming you on a unrelated issue, would not be better to load snippets based on filetype just like the way luasnip [does](https://github.com/L3MON4D3/LuaSnip/blob/master/lua/luasnip/loaders/from_vscode.lua#L212)

**Comment by hrsh7th** [comment]
Improve vsnip performance with friendly-snippets. 
https://github.com/hrsh7th/vim-vsnip/commit/87d144b7451deb3ab55f1a3e3c5124cfab2b02fa

**Comment by mortezadadgar** [comment]
> Improve vsnip performance with friendly-snippets.
> [hrsh7th/vim-vsnip@87d144b](https://github.com/hrsh7th/vim-vsnip/commit/87d144b7451deb3ab55f1a3e3c5124cfab2b02fa)

Thanks! it's much faster now
